jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -266,139 +265,144 @@ public String toString()
         return String.format("%s@%x{s=%d/%d,p=%d}",
             getClass().getSimpleName(),
             hashCode(),
-            _size.get(),
+            _count.getLo(),
             _capacity,
-            _pending.get());
+            _count.getHi());
     }
 
-    private class ReservedThread implements Runnable
+    private enum State
     {
-        private final SynchronousQueue<Runnable> _task = new SynchronousQueue<>();
-        private boolean _starting = true;
-
-        public boolean offer(Runnable task)
-        {
-            if (LOG.isDebugEnabled())
-                LOG.debug("{} offer {}", this, task);
-
-            try
-            {
-                _task.put(task);
-                return true;
-            }
-            catch (Throwable e)
-            {
-                LOG.ignore(e);
-                _size.getAndIncrement();
-                _stack.offerFirst(this);
-                return false;
-            }
-        }
+        PENDING,
+        RESERVED,
+        RUNNING,
+        IDLE,
+        STOPPED
+    }
 
-        public void stop()
-        {
-            offer(STOP);
-        }
+    private class ReservedThread implements Runnable
+    {
+        // The state and thread are kept only for dumping
+        private volatile State _state = State.PENDING;
+        private volatile Thread _thread;
 
         private Runnable reservedWait()
         {
             if (LOG.isDebugEnabled())
-                LOG.debug("{} waiting", this);
+                LOG.debug("{} waiting {}", this, ReservedThreadExecutor.this);
 
             while (true)
             {
                 try
                 {
-                    Runnable task = _idleTime <= 0 ? _task.take() : _task.poll(_idleTime, _idleTimeUnit);
+                    // Always poll at some period as safety to ensure we don't poll forever.
+                    Runnable task = _idleTime <= 0 ? _queue.poll(30, TimeUnit.SECONDS) : _queue.poll(_idleTime, _idleTimeUnit);
```

Comment:
The hardcoded 30 seconds should be either the default idle timeout (1 minute) or a much shorter time like 1 second to speed up stopping.

If we stop while entering this method, this thread is going to wait 30 seconds (the offer of the STOP task will fail, as the thread is still not polling), which seems like a long time.
Waiting just 1 second will make the stop edge case faster, and if the server is idle, it's a loop around every second.